Zend/zend_attributes.c

```diff
@@ -1,21 +1,62 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Benjamin Eberlei <kontakt@beberlei.de>                      |
+   |          Martin Schröder <m.schroeder2007@gmail.com>                 |
+   +----------------------------------------------------------------------+
+*/
+
 #include "zend.h"
 #include "zend_API.h"
 #include "zend_attributes.h"
+#include "zend_attributes_arginfo.h"
 
-ZEND_API zend_class_entry *zend_ce_php_attribute;
+ZEND_API zend_class_entry *zend_ce_attribute;
 
-static HashTable internal_validators;
+static HashTable internal_attributes;
 
-void zend_attribute_validate_phpattribute(zend_attribute *attr, int target)
+void validate_attribute(zend_attribute *attr, uint32_t target, zend_class_entry *scope)
 {
-	if (target != ZEND_ATTRIBUTE_TARGET_CLASS) {
-		zend_error(E_COMPILE_ERROR, "Only classes can be marked with <<PhpAttribute>>");
+	if (attr->argc > 0) {
+		zval flags;
+
+		if (FAILURE == zend_get_attribute_value(&flags, attr, 0, scope)) {
+			return;
+		}
+
+		if (Z_TYPE(flags) != IS_LONG) {
+			zend_error_noreturn(E_ERROR, "Flags must be an integer value, %s given", zend_zval_type_name(&flags));
```

Comment:
```suggestion
			zend_error_noreturn(E_ERROR, "Attribute::__construct(): Argument #1 ($flags) must must be of type int, %s given", zend_zval_type_name(&flags));
```
would be more typical.